app/ide-desktop/lib/dashboard/src/layouts/AssetVersions/AssetVersions.tsx

```diff
@@ -23,56 +39,108 @@ import type AssetTreeNode from '#/utilities/AssetTreeNode'
 /** Props for a {@link AssetVersions}. */
 export interface AssetVersionsProps {
   readonly item: AssetTreeNode
+  readonly dispatchAssetListEvent: (event: assetListEvent.AssetListEvent) => void
 }
 
 /** A list of previous versions of an asset. */
 export default function AssetVersions(props: AssetVersionsProps) {
-  const { item } = props
+  const { item, dispatchAssetListEvent } = props
   const { backend } = backendProvider.useBackend()
   const { getText } = textProvider.useText()
   const toastAndLog = toastAndLogHooks.useToastAndLog()
+  const [placeholderVersions, setPlaceholderVersions] = React.useState<
+    readonly backendService.S3ObjectVersion[]
+  >([])
   const isCloud = backend.type === backendService.BackendType.remote
-
-  const {
-    status,
-    data: versions,
-    isPending,
-  } = useAssetVersions.useAssetVersions({
+  const queryKey = ['assetVersions', item.item.id, item.item.title]
+  const versionsQuery = useAssetVersions.useAssetVersions({
     backend,
+    queryKey,
     assetId: item.item.id,
     title: item.item.title,
     onError: backendError => toastAndLog('listVersionsError', backendError),
     enabled: isCloud,
   })
+  const latestVersion = versionsQuery.data?.find(version => version.isLatest)
 
-  const latestVersion = versions?.find(version => version.isLatest)
+  const restoreMutation = reactQuery.useMutation({
+    mutationFn: async (variables: AddNewVersionVariables) => {
+      if (item.item.type === backendService.AssetType.project) {
+        await backend.restoreProject(item.item.id, variables.versionId, item.item.title)
+      }
+    },
+    onMutate: variables => {
+      setPlaceholderVersions(oldVersions => [
+        {
+          isLatest: false,
+          key: uniqueString.uniqueString(),
+          lastModified: dateTime.toRfc3339(new Date()),
+          versionId: variables.placeholderId,
+        },
+        ...oldVersions,
+      ])
+    },
+    onSuccess: async () => {
+      // `backend.restoreProject` does not return the ID of the new version, so a full refetch is
+      // necessary.
+      await versionsQuery.refetch()
+    },
+    onError: (error: unknown) => {
+      toastAndLog('restoreProjectError', error, item.item.title)
```

Comment:
Not sure it's a good idea to toast error, because it will frustrate users - error's happened, but can't do anything with the error. Maybe we can show additional actions in toast? Or maybe display error inline?

Reply:
i don't know about additional actions in toast tbh. inline sounds fine but not sure what's the best way to send the error info back to the button that was responsible for the error